go/libraries/doltcore/doltdb/table_edit_session.go

```diff
@@ -0,0 +1,239 @@
+// Copyright 2020 Liquidata, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package doltdb
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/liquidata-inc/dolt/go/libraries/doltcore/schema"
+)
+
+// TableEditSession handles all edit operations on a table that may also update other tables. Serves as coordination
+// for SessionedTableEditors.
+type TableEditSession struct {
+	Props TableEditSessionProps
+
+	root       *RootValue
+	tables     map[string]*SessionedTableEditor
+	writeMutex *sync.RWMutex // This mutex is specifically for changes that affect the TES or all STEs
+}
+
+// TableEditSessionProps are properties that define different functionality for the TableEditSession.
+type TableEditSessionProps struct {
+	ForeignKeyChecksDisabled bool // If disabled, then all foreign key checks and updates are ignored
+}
+
+// CreateTableEditSession creates and returns a TableEditSession. Inserting a nil root is not an error, as there are
+// locations that do not have a root at the time of this call. However, a root must be set through SetRoot before any
+// table editors are returned.
+func CreateTableEditSession(root *RootValue, props TableEditSessionProps) *TableEditSession {
+	return &TableEditSession{
+		Props:      props,
+		root:       root,
+		tables:     make(map[string]*SessionedTableEditor),
+		writeMutex: &sync.RWMutex{},
+	}
+}
+
+// GetTableEditor returns a SessionedTableEditor for the given table. If a schema is provided and it does not match the one
+// that is used for currently open editors (if any), then those editors will reload the table from the root.
+func (tes *TableEditSession) GetTableEditor(ctx context.Context, tableName string, tableSch schema.Schema) (*SessionedTableEditor, error) {
+	tes.writeMutex.Lock()
+	defer tes.writeMutex.Unlock()
+
+	return tes.getTableEditor(ctx, tableName, tableSch)
+}
+
+// GetRoot returns an updated root with all of the changed tables.
+func (tes *TableEditSession) GetRoot(ctx context.Context) (*RootValue, error) {
+	tes.writeMutex.Lock()
+	defer tes.writeMutex.Unlock()
+
+	rootMutex := &sync.Mutex{}
+	wg := &sync.WaitGroup{}
+	wg.Add(len(tes.tables))
+
+	newRoot := tes.root
+	var tableErr error
+	var rootErr error
+	for tableName, ste := range tes.tables {
```

Comment:
So a table editor, for the most part, is only concerned with itself. Let's say you have two tables, A and B, and two separate functions grab an editor for one of them. That editor is used just like the old editor, in that you issue your Insert, Update, etc. to them. Now, let's say that one of the editors is done and wants to grab the root, that's when they call this. That should have no observable effect on the other editor. The only time an editor should be deleted is if the table is deleted, otherwise you have to synchronize your table editors.

The goal of the `TableEditSession` is to track all of the internal changes for you. You, as a caller who only wants to edit or update a single table, should not be concerned with any other user of this edit session.

Reply:
I think this makes sense, although I do still worry we'll get ourselves in trouble with this. But agree that letting clients have references to the table editors makes it necessary to keep them around. Fixing that (requiring all edits to go through the session) would be a lot of work and would have some drawbacks.